File: src/main/java/com/example/demoAllTouhou/services/PowersService.java
package com.example.demoAllTouhou.services;

import com.example.demoAllTouhou.entities.PowersEntity;
import com.example.demoAllTouhou.repositories.PowersRepository;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class PowersService {

    private final PowersRepository powersRepository;

    public PowersService(PowersRepository powersRepository) {
        this.powersRepository = powersRepository;
    }

    // **Crear un nuevo Power**
    public PowersEntity createPower(PowersEntity powersEntity) {
        return powersRepository.save(powersEntity);
    }

    // **Obtener todos los Powers**
    public List<PowersEntity> getAllPowers() {
        return powersRepository.findAll();
    }

    // **Obtener un Power por ID**
    public PowersEntity getPowerById(Long id) {
        return powersRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Power not found with id: " + id));
    }

    // **Actualizar un Power**
    public PowersEntity updatePower(Long id, PowersEntity updatedPower) {
        PowersEntity existingPower = powersRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Power not found with id: " + id));

        if (updatedPower.getName() !=null && !updatedPower.getName().isEmpty()) existingPower.setName(updatedPower.getName());
        if (updatedPower.getDescription() !=null && !updatedPower.getDescription().isEmpty()) existingPower.setDescription(updatedPower.getDescription());

        return powersRepository.save(existingPower);
    }

    // **Eliminar un Power por ID**
    public void deletePower(Long id) {
        if (!powersRepository.existsById(id)) {
            throw new RuntimeException("Power not found with id: " + id);
        }
        powersRepository.deleteById(id);
    }

    // **Obtener Powers por Name (contiene string)**
    public List<PowersEntity> getPowersByName(String name) {
        return powersRepository.findByNameContainingIgnoreCase(name);
    }

    // **Obtener Powers por Character ID**
    public List<PowersEntity> getPowersByCharacterId(long characterId) {
        return powersRepository.findByCharacters_Id(characterId);
    }
}

File: Frontend/src/services/PowersService.js
import httpClient from "../http-common";

// Crea un nuevo Power
const createPower = (power) => {
    return httpClient.post(`/powers/create`, power);
};

// Obtiene todos los Powers
const getAllPowers = () => {
    return httpClient.get(`/powers/all`);
};

// Obtiene un Power por ID
const getPowerById = (id) => {
    return httpClient.get(`/powers/${id}`);
};

// Actualiza un Power
const updatePower = (id, updatedPower) => {
    return httpClient.put(`/powers/${id}`, updatedPower);
};

// Elimina un Power por ID
const deletePower = (id) => {
    return httpClient.delete(`/powers/${id}`);
};

// Obtiene Powers por Name (contiene string)
const getPowersByName = (name) => {
    return httpClient.get(`/powers/search/name`, { params: { name } });
};

// Obtiene Powers por Character ID
const getPowersByCharacterId = (characterId) => {
    return httpClient.get(`/powers/character/${characterId}`);
};

// Obtiene múltiples Powers por lista de IDs
const getPowersByIds = (idList) => {
    return httpClient.post(`/powers/by-ids`, idList);
};

export default {
    createPower,
    getAllPowers,
    getPowerById,
    updatePower,
    deletePower,
    getPowersByName,
    getPowersByCharacterId,
    getPowersByIds
};
